python/ray/ml/train/trainer.py

```diff
@@ -0,0 +1,71 @@
+import abc
+import logging
+from typing import Dict, Union, Callable, Optional, TYPE_CHECKING
+
+from ray.ml.preprocessor import Preprocessor
+from ray.ml.checkpoint import Checkpoint
+from ray.ml.result import Result
+from ray.ml.config import ScalingConfig, RunConfig
+from ray.tune.trainable import ConvertibleToTrainable
+from ray.util import PublicAPI
+
+if TYPE_CHECKING:
+    from ray.data import Dataset
+
+# A type representing either a ray.data.Dataset or a function that returns a
+# ray.data.Dataset and accepts no arguments.
+GenDataset = Union["Dataset", Callable[[], "Dataset"]]
+
+
+logger = logging.getLogger(__name__)
+
+
+@PublicAPI(stability="alpha")
+class TrainingFailedError(RuntimeError):
+    """An error indicating that training has failed."""
+
+    pass
+
+
+@PublicAPI(stability="alpha")
+class Trainer(ConvertibleToTrainable, abc.ABC):
```

Comment:
happy to disagree and commit here, but I'd like to hear @ericl and @richardliaw 's view on why not having an _explicit_ `as_trainable` one-liner method here. If I'm looking up the Trainer interface I want to readily see what I have to implement.

Reply:
Good point. Amog, Is convertibletotrainable interface used otherwise?

If yes and often so, then I think the benefit of reuse here outweighs the readability. My take is that on RTD it will render in a user friendly manner, so the downsides is limited.

If not then I am for collapsing it into 1 interface.

Reply:
I don't think it's being used anywhere else but the `Trainer`. Is that right @krfricke @xwjiang2010?

Reply:
Addressed offline.